### ISSUE TITLE
relax requirement of NIP-42 to include the relay

### ISSUE THREAD

**Issue by mleku** [nostr-protocol/nips#1013]
the only things that are important about auth are challenge, pubkey and signature, the relay is not relevant

the requirement of the relay stops projects like bostr from being able to proxy your auth because there is no protocol with the client to accept a different relay address from an IP

the relay address can be fake anyway, it is not meaningful or anything related to security

by dropping the requirement of auth events to contain a relay tag, proxy relays become possible and this allows a relay to broadcast your events to your relay list - all automatically, and on the VPS running it, all that is required then is your pubkey as an access control whitelist, and a little more code in the proxy relay and you can fully proxy, and the proxy can also then do other things like caching your requests so it doesn't pester the other relays for them when it has them in cache

i am going to remove the restriction in my relay where it doesn't accept auth events with matching challenges because of the relay tag, or even the absence of the relay tag, only enforcing the challenge and the time window

the challenge is all of the security, and the relay tag just makes the auth protocol inflexible

**Comment by vitorpamplona** [comment]
Ack. Nice.

**Comment by Yonle** [comment]
Strongly disagree on this proposal.

This will give numerous of privacy issues (even in a small amount), for example there are some relays that is just another proxy of an nostr relay that all it does was forwarding any AUTH request from upstream to client in order to record who's the users are PMing to.

```
Victim <-> MITM Relay Proxy <-> Upstream
```

Of course the MITM relay proxy will not be able to read your PM content due to it being encrypted, However the MITM will know who are you talking to, which what some paid relays want to avoid to.

**Comment by staab** [comment]
> we should require the relay address to be in the auth challenge

I agree, a client really needs to know who is asking for auth so they can decide whether to authenticate. Currently it's impossible for clients to know who they're authenticating with; adding the relay url to the challenge would allow clients to compare the challenge url with the connection url and decide what happens if they're different.

I'm currently working on a relay proxy as it happens, and the MITM attack is very real. If the proxy's connection with the destination relay is authenticated, they can hijack an authenticated session and exfiltrate whatever data they want, or publish unwanted events to the destination relay. This is true regardless of the format of the AUTH message. Proxies _must_ be trusted, which is hard if clients connect to whomever and authenticate freely. User-controlled white/black lists of relays is probably what's ultimately needed.

**Comment by mikedilger** [comment]
I argued for a nonce in the AUTH protocol but it was discarded because replay was supposedly prevented via other means (I think by adding the relay).  I'm still skeptical of AUTH (it feels far too simple to be correct) but I can't explain how to break it.

I'm not sure how MITM comes about. Clients start the websocket connection. How does that end up connecting to a different machine than the one they are trying to connect to?  DNS poisoning?  How does the SSL get spoofed?  I think it takes a nation state with control of top level CAs to do that.

**Comment by vitorpamplona** [comment]
I didn't realize this was a form of attack, so I am changing my mind about removing the tag. 

The `relay` tag prevents one of your current relays from going evil and connecting to every other relay to do authenticated things as you. Without the `relay`, they can collect all challenges, send it to you, get the signed events back, and forward it to the next relay. Their connection to that relay is now authenticated as you. 

In that case, public relays can then have access to private relays and download/index all events from over there. 

That being said, the tag does not avoid MITM as @mikedilger alluded to. Governments control CAs, DNS, and IP layers and can place any proxy server in between you and your real relay with the same DNS and even IP that you are used to. These types of attacks were one of our main concerns when we were exchanging health information between two countries at war with one another. Everything is possible when the military gets involved.

**Comment by staab** [comment]
I'm starting to wonder if it's just not worth it to have authenticated relay proxies built into the protocol. Currently, the protocol makes it impossible — maybe that's a good thing, because establishing proxy trust is basically impossible without self-hosting. For that case, it would be easy to self-host a relay, give it your private key, and have it authenticate on your behalf.

**Comment by staab** [comment]
Drafted PR #1016 to see what this looks like. Most of the PR is warnings about MITM. Maybe we should leave this one alone.

**Comment by mleku** [comment]
a couple of points

- only the relay has a websocket with the matching auth challenge, there is nothing about the auth response having the relay tag in it that changes that
- a proxy relay will gain the ability to make requests without knowledge of the user, and this is because requests are not signed - this could be solved by wrapping the requests in an event, putting them into its content field
- a proxy relay will not be able to publish events in the name of the user because they cannot sign as that user, this is a hole that can be closed just by making it so authed users in general cannot publish events that don't match their pubkey
- the proxy is going to see all your DMs anyway, and nothing stopping them from passively capturing the metadata out of them, the only way around this is an actual VPN like wireguard
- the use case of a proxy relay is mostly for people running their own so trust requirement is low, since the user can control what software they use
- to make it into a sellable service there has to be a whole secondary layer of encryption that hides the traffic from the proxy
